Fixed #37111 -- Added sprints quickstart docs page. (#21338)
django/django: docs/internals/contributing/sprint-quickstart.txt

=======================
Sprint Quickstart Guide
=======================

This document outlines the concise guidelines for preparing to contribute to
Django. This should be used during sprints or by people who are more familiar
with contributing to Python open source projects.

This guide assumes you're familiar with:

* Git

* Managing multiple versions of Python

* Python dependency management, including virtual environments

* The concept of running tests and building documentation

Set up
======

Fork the repository, clone your fork to your machine and configure a remote
reference to Django's repository.

.. code-block:: shell

    $ git clone https://github.com/YourGitHubName/django.git django-repo
    $ cd django-repo
    $ git remote add upstream https://github.com/django/django.git

> Install the dependencies to run the tests and build the documentation.

Reviewer: ```suggestion
Install the dependencies to run the tests and build the documentation. Make sure you have a Python version >=3.12 installed.
```
The check for a Python version >=3.12 could go here.

Author: ```suggestion
Install the dependencies to run the tests and build the documentation.
Django requires you to use Python version 3.12 or greater.
```

Author: I revised this line in: https://github.com/django/django/pull/21338/commits/176acbb1667ca927823ba3c2717c1e304db1b6bb

.. code-block:: shell

>     $ python -m venv .venv

Reviewer: ```suggestion
    $ python3.14 -m venv .venv  # Replace 3.14 with your Python version >=3.12
```
Then the venv command could contain a specific Python version (in case of parallel Python installations).

Author: I think this may be problematic for windows as I don't think versioned executables are installed. It may also be a problem for people using the default python, but I suspect that may not be 3.12 so it's fine it's a problem. Not sure. I'd like others' thoughts here too.

Author: This is also one of the assumptions we're making up top.

> * Managing multiple versions of Python

    $ source .venv/bin/activate
    $ python -m pip install -e .
    $ python -m pip install -r tests/requirements/py3.txt
    $ python -m pip install -r docs/requirements.txt
    $ # This needs to be >= 3.12

>     $ python -V

Author: ```suggestion
    $ # This needs to be >= 3.12
    $ python -V
    $ python -m venv .venv
    $ source .venv/bin/activate
    $ python -m pip install -e .
    $ python -m pip install -r tests/requirements/py3.txt
    $ python -m pip install -r docs/requirements.txt
```

> Confirm that the version of Python you're using is Python 3.12 or greater.

Reviewer 2: Could these lines go earlier on? (Since by this point it's too late to check for the version if you're not on 3.12.)

Author: Moving this discussion up to https://github.com/django/django/pull/21338/files#r3299978260

Run the tests:

.. code-block:: shell

    $ ./tests/runtests.py

> If ten or fewer tests fail, you're likely okay to move on. However, keep in

Reviewer: It might be helpful for new contributors to know why tests may fail and why that may be okay. They might not be familiar with the concept of flaky tests.

Author: I revised this language in https://github.com/django/django/pull/21338/commits/346842a362ec62ef163155df8dbe1352fd1e778e. I didn't explain the concept of flakiness, but did provide more context around why some may fail based on Jacob's comments in this thread: https://github.com/django/django/pull/21338/files#r3325874401

mind if the code you're changing requires changing those tests, you'll need

> to fix the underlying issues on your machine first. There are 10,000+ tests

Reviewer 2: ```suggestion
If ten or fewer tests fail, you're likely okay to move on. However, keep in
mind if the code you're changing requires changing those failing tests, you'll need
to fix the underlying issues on your machine first. There are 10,000+ tests
```
(I think this sentence makes sense already, but I did have to read it twice to clock it's meaning. I wondered if being more explicit about what "those" meant, might improve clarity? Perhaps this is just me though, so feel free to ignore.)

though, so it's unlikely.

Build the documentation:

.. code-block:: shell

    $ (cd docs && make html)

Finding an issue to work on
===========================

> Finding a workable ticket is a challenging issue. The ways to get started

Reviewer 2: ```suggestion
Finding a workable ticket is a challenging issue. Some ways to get started
```

are the following:

* Look at the `easy pickings`_.

* Use the `vulture method`_.

The vulture method
------------------

This approach to finding tickets is to look for accepted tickets that
haven't been touched in six months or longer. The ideal ticket has a PR
attached to it that has a code review with explicit requests and the author
has not responded in at least six months.

When you find a stale ticket, make yourself the owner on the ticket with
the "Modify Ticket" button near the bottom of the page. Claim the ticket
by clicking the "assign to" radio button in the "Action" section. Your
username will be filled in the text box by default.

Assessing issues
----------------

> 

Reviewer 2: I wonder if it's worth having these two paragraphs before the discussion of easy-pickings and the vulture method. They contain really valuable guidance that applies regardless of the approach you use, and I think it would be worth a new contributor reading them before ploughing ahead.

Author: I would like to think that a person would read this entire section before moving forward with selecting an issue. It's not long there isn't code that would suggest a 1, 2, 3, step process like we have above with the code set up.

That said, it's easy enough to explain how to assess an issue, then how to find the easier ones. I think that would flow well too.

Author: I reworked the section, including replacing the word "issue" with "ticket" where applicable. I like it less because we don't have the linked header on vulture method so linking directly is harder, but I think as a whole it flows better.

Reviewer 2: Hmm... I see what you mean about link-ability. But I do agree there is a more logical flow this way round. I suppose you can still link to the "methods" section, which is mostly about the vulture method.

Author: Can we resolve this conversation?

Reviewer 2: We can indeed. Although I'm on my phone only for the next 18 hours and can't figure out how to do that on the mobile UI so I'll let you do the honors.

When looking over tickets, it's a good idea to keep your search to
tickets that were created in the last five years. An easy way to tell is
by the ticket number. Avoid tickets with an ID of less than 20,000. These
are old for a reason. You're welcome to try, but they will require a high
degree of perseverance.

Consider using the component filter option in Trac to limit your search
to areas you have used before. If you've used forms in a few Django apps,
consider filtering down to the forms components. Similarly, if you've
never worked with geospatial projects, it's probably wise to avoid
``django.contrib.gis``.

> 

Reviewer 2: I've seen a couple of github review-comments recently reminding folks to credit the authors they have "scavenged" from (to extend the vulture metaphor). Could we put a friendly reminder here? We do mention crediting others elsewhere in the docs, but not explicitly for this kind of approach.

Author: I think adding a line at the end of the section that reminds them and links them to our proper documentation would be fine.

The reason I think it should be 0 or 1 lines, is that this document is what a person needs to get started. Not what they need to do everything. Cleaning up a commit so it's ready for check-in is a few steps past what the scope of this document.

Reviewer 2: Yes, I had similar thoughts since this docs is minimal in scope. I think a better solution might be to beef up (only slightly) [this section](https://github.com/django/django/blob/c0b12e190beacc9b458f40abec7a3f8963ade9fc/docs/internals/contributing/committing-code.txt#L141-L142)  and then add a few words here which link to it.

Author: I added a sentence to the linked section clearly stating they should credit the previous authors and then linked to it like you suggested in https://github.com/django/django/pull/21338/commits/9a7e35b5a5e03e6d21458ccc038e9d20f53a0e33 Thanks for calling that out. Let me know if you think there's something else we can improve.

Reviewer 2: That looks good to me :)

=======================
Sprint Quickstart Guide
=======================

This document outlines the concise guidelines for preparing to contribute to
Django. This should be used during sprints or by people who are more familiar
with contributing to Python open source projects.

This guide assumes you're familiar with:

> * Git

Reviewer 3: Shall we offer more detailed resources when we have one?
```suggestion
* Git (see :doc:`how-to use Git </internals/contributing/writing-code/working-with-git>`)
```

* Managing multiple versions of Python

* Python dependency management, including virtual environments

* The concept of running tests and building documentation

Set up
======

Fork the repository, clone your fork to your machine and configure a remote
reference to Django's repository.

> .. code-block:: shell

Reviewer 3: Django contributing docs state that for console code we should use `.. console::`. Is there a reason not to do that in this file?

```suggestion
.. console::
```

Author: I'm not sure if there's a good reason to do this one way or another. I copied from the testing.txt file which uses `code-block:: shell`

    $ git clone https://github.com/YourGitHubName/django.git django-repo
    $ cd django-repo

>     $ git remote add upstream https://github.com/django/django.git

Reviewer 3: My preferred advice for cloning for contributing is the following, it has the advantage that pulling `main` brings `upstream`'s:

```suggestion
    $ git clone https://github.com/django/django.git django-repo
    $ cd django-repo
    $ git remote rename origin upstream
    $ git add remote origin git@github.com:YourGitHubName/django.git
```

Author: This was copied from unit-tests.txt. Should we be changing that as well for consistency?

Reviewer 4: Given that, sounds out of scope to me.

> it has the advantage that pulling main brings upstream's

I've never seen this implicit behavior before, I prefer the explicitness of pulling upstream.

Reviewer 3: @jacobtylerwalls there is nothing implicit, it's just that you dont need to update your fork's main to get latest upstream's main. In any case, no worries I'm happy to leave as is!

Install the dependencies to run the tests and build the documentation.

.. code-block:: shell

    $ python -m venv .venv
    $ source .venv/bin/activate
    $ python -m pip install -e .
    $ python -m pip install -r tests/requirements/py3.txt

>     $ python -m pip install -r docs/requirements.txt

Reviewer 3: What's the motivation to prefer `python -m pip install ...` vs `pip install ... `? (really curious)

Also I find it more useful to install all in single call (less cognitive load), and let's not forget to upgrade `pip`:

```suggestion
    $ pip install -e . -U pip -r tests/requirements/py3.txt -r docs/requirements.txt
```

Author: I never remember these things can be combined :sweat_smile:

Reviewer 4: > What's the motivation to prefer python -m pip install

We settled on the `-m` flavor in aed89adad54a977829c4f180b036033e031ebcc7. I think we should stick to it to avoid getting bikesheddy followups.

> (really curious)

This was a whole meme a few years ago. In brief, if your `$PATH` is even slightly unexpected, you can resolve to a binary outside your venv. @nessita you saw that happen to me recently with `twine`, where my homebrew python was unexpected first in my path. (Despite all my efforts to avoid homebrew pythons, I think one was pulled in after installing postgres.app or something.)

Reviewer 3: Fair enough! settling on one command line using `python -m pip install ...` -- resolving!

>     $ # This needs to be >= 3.12

Reviewer 3: What's our strategy to ensure this hard coded value gets updated when the minimum supported Python version is bumped?

Author: In my experience, a person who bumps the minimum supported version does a search for "3.10" and audits all usages to see what needs to change. I'd expect this to be caught in that process.

Reviewer 2: Could we define something in the sphinx `conf.py` alongside `version` so it would be impossible to miss when updating, e.g. a variable called `minimum_supported_python_version`. We could stick it in the `rst_epilog` so we could use it in the docs where needed. 

I couldn't find many places we'd need it, so maybe it would be overkill, but I did see a similar issue in the tutorial:
https://github.com/django/django/blob/c0b12e190beacc9b458f40abec7a3f8963ade9fc/docs/intro/tutorial01.txt#L26

Reviewer 4: That seems like overkill to me. I like Tim S.'s suggestion to leave this as is.

Reviewer 3: I'm happy to leave it like this in this PR, but I do think automating the Python version used in these examples is worthwhile. We've had other tickets about bumping the Django version used in examples elsewhere in the contributing docs, and while each individual update is small and straightforward, the cumulative cost of ticket triage, review, backports, and releases adds up. In my view, that cost outweighs the effort of automating these version bumps.

    $ python -V

> Confirm that the version of Python you're using is Python 3.12 or greater.

Reviewer 3: Could we perhaps be future-proof and say the following, I would also move this before any python invocation:

```suggestion
Confirm that the version of Python you're using is supported for the most recent version of Django (list of supported version is at :ref:`faq-python-version-support`):

    $ python -V

```

Author: I think the purpose of this document is to avoid having to open other pages to use the instructions. Having them go to another page to see if they have something set up correctly goes against that.

Author: I moved the `python -V` up in this commit: https://github.com/django/django/pull/21338/commits/176acbb1667ca927823ba3c2717c1e304db1b6bb

We can discuss the 3.12 change in this thread: https://github.com/django/django/pull/21338/files#r3320099578

Run the tests:

.. code-block:: shell

    $ ./tests/runtests.py

> If ten or fewer tests fail, you're likely okay to move on. However, keep in

Reviewer 3: I'm not very comfortable documenting this, the Django docs explicitly say "It's our policy to make sure all tests pass at all times." (from https://github.com/django/django/blob/main/docs/internals/contributing/writing-code/unit-tests.txt#L6C12-L6C69)

Author: I started this document as a reproduction of exactly what I was telling people as a starting point. Today, there is a reasonable chance that a person going through the set up and running all the tests will have a local failure. I don't know the percentage, but it happens enough on the contributing-getting-started channel that I knew to just tell people at sprints to ignore when few failures occurred.

This seems reasonable because contributors want to feel useful. Spending extra time getting their machine configured and not working on a tangible issue is an energy vampire. When a person is at sprints I believe this is even more important.

Admittedly, the easy answer is to remove this and say "all tests will pass" and expect people to realize there's a nuance implied. If there's value in communicating that nuance, I'm happy to iterate on this.

Reviewer 4: I think we can adjust the wording to avoid making it sound like "we're okay with failures," but I agree with the need for briefly explaining, "if you have just a handful of failures, your smoke test for your setup has passed. (You can research whether or not the failing tests are known issues if you have an interest.)"

I was elbow-side with someone at Django on the Med, and they drained a lot of time re-investigating little edge cases in `test_strip_tags` that varied by Python point release. Despite our "we're perfect and our tests always pass" language, we also disclaim any support below the latest Python point release. I've merged cleanups that I know will cause test failures on lower point releases of Python 3.12.

I don't know that we need to force a sprint contributor to reinstall the latest point release if they don't feel like it. It's just a sprint.

Author: I revised the language in https://github.com/django/django/pull/21338/commits/346842a362ec62ef163155df8dbe1352fd1e778e I think it makes it clear that the tests do all pass, but acknowledges that people's local machines are a different thing.

Reviewer 3: I like the latest proposal, with the request of instead of duplicating the full django-docker-box URL, we should link (using an internal reference) to the unit testing page where it's already mentioned. Duplicating absolute URLs increases maintenance cost by creating multiple places that need updating if the target ever changes (which does happen sadly).

Author: Changed to your request in 4bc0f1a1a35baf62b12e82268e2a120486d95467

mind if the code you're changing requires changing those failing tests, you'll need
to fix the underlying issues on your machine first. There are 10,000+ tests
though, so it's unlikely.

Build the documentation:

.. code-block:: shell

>     $ (cd docs && make html)

Reviewer 3: Parenthesis are unnecessary here, and I would dare to say, confusing.

If for any reason the reader copies the line in a way that the space between `$` and `(` is removed, `$(...)` is a bash command substitution syntax so the shell would try to execute the output of the subshell as a command, producing a confusing error rather than building the docs. My advice:

```suggestion
    $ cd docs && make html
```

Author: For context, I did this because the parenthesis keep you in the `/django` directory. Removing them leaves you in `/django/docs`. I suppose we could chain another ` && cd .. ` to resolve that.

Author: ```suggestion
    $ cd docs && make html && cd ..
```

Finding a ticket to work on
===========================

Finding a workable ticket is a challenging task. It's important

> to read this whole section before attempting to find a ticket to work on.

Reviewer 3: All the paragraphs should wrap at 80 cols (including newline) and should use as much of the 80 cols as possible. I can push these cleanups later on if that helps.

Author: Yeah, I gave up on bothering with it until we have things improved. Trying to respect this style guideline makes suggestion diffs less helpful and code reviews harder.

Assessing tickets
-----------------

When looking over tickets, it's a good idea to keep your search to
tickets that were created in the last five years. An easy way to tell is
by the ticket number. Avoid tickets with an ID of less than 20,000. These
are old for a reason. You're welcome to try, but they will require a high

> degree of perseverance.

Reviewer 3: I think this advice is great, but I wouldn't use exact ticket number range because this does not age well. I would suggest linking the users to perhaps not start with tickets "modified before than N years".

Author: I was going to explain why using the ID is worth it, but instead can we change the default columns on Trac to include created and modified? That way we can say tickets "created in the last N years".

Edit: I don't think new contributors are going to realize they can use the add column feature of Trac. I always forget about it.

Author: You can ignore my ask about setting the default. I added a link with the customized query in https://github.com/django/django/pull/21338/commits/df1d0be3982bd69fcf260fbf16e2b1813c5c1227 I'll switch out the N years next.

Author: Resolve the N years reference in https://github.com/django/django/pull/21338/commits/90bc97ed20c1a8b4da40c58015069a073ee80a8a

Reviewer 3: This is better, thank you! I think we can polish this section a bit more. Two things:

1. Since we removed the ">20000" reference, "An easy way to tell is by the ticket number" no longer tells the reader anything unless they already know ticket numbers are sequential. The paragraph also gives three thresholds in a row: "last five years", then "avoid older than 10 years", then "last five years" again. We should pick one and state it once.

2. The two new query links only added `col=time`, which displays the Created column but doesn't filter on it. So they return tickets of any age, which undercuts the five-year advice the prose just gave. The vulture link already uses natural-language dates, so we can do the same here and have the links actually deliver what the text promises.

Combining both, plus folding the bullets into the prose so the documentation query reads as a concrete example of the component advice:

```rst
When looking over tickets, keep your search to `accepted tickets created in
the last five years`_. Older tickets are old for a reason: they are usually
still open because they are hard, contentious, or both. You're welcome to
try one, but expect it to require a high degree of perseverance.

Consider using the component filter option in Trac to limit your search to
areas you have used before. If you've used forms in a few Django apps,
consider filtering down to the forms component. Similarly, if you've never
worked with geospatial projects, it's probably wise to avoid
``django.contrib.gis``. For example, here are the `tickets for the
documentation component`_.

.. _accepted tickets created in the last five years: https://code.djangoproject.com/query?stage=Accepted&status=assigned&status=new&time=5+years+ago..today&order=id&desc=1&col=id&col=summary&col=owner&col=type&col=component&col=time&col=changetime
.. _tickets for the documentation component: https://code.djangoproject.com/query?component=Documentation&stage=Accepted&status=assigned&status=new&time=5+years+ago..today&order=id&desc=1&col=id&col=summary&col=owner&col=type&col=component&col=time&col=changetime
```

I have this changes locally since I've been iterating over and over, I can push this unless you object.

Author: Generally those are good changes, but I would not apply the language of "they are usually still open because they are hard, contentious, or both" to things in the 5-9 years old range though. When I wrote that line, it was meant for things that are legit old, not things that just slip through the cracks year after year.

Author: @nessita I do feel that there's a difference between a 10+ year old ticket and something younger, but I'll defer to you on this.

Reviewer 3: I'll push this and restore the 10+ years per your preference.

consider filtering down to the forms components. Similarly, if you've
never worked with geospatial projects, it's probably wise to avoid
``django.contrib.gis``.

Methods to find tickets
-----------------------

Some ways to find a ticket to work on are the following:

* Look at the `easy pickings`_.

* Use the `vulture method`_.

The vulture method is the process where you look for accepted tickets that
haven't been touched in six months or longer. The ideal ticket has a PR
attached to it that has a code review with explicit requests and the author
has not responded in at least six months.

When you find a stale ticket, make yourself the owner on the ticket with
the "Modify Ticket" button near the bottom of the page. Claim the ticket
by clicking the "assign to" radio button in the "Action" section. Your
username will be filled in the text box by default.

Resuming work from someone else's branch
========================================

If you're continuing the work from someone else's pull request on GitHub,
things can be tricky. You need to pull their branch, but then push it to
your GitHub repository fork.

> Run the following, but switch ``ThePRNumber`` with the **pull request ID**

Reviewer 3: I was quite confused for a while thinking what is a PR ID that I don't know about, but then I figured you meant number by ID? shall we say that to avoid possible confusions?

```suggestion
Run the following, but replace ``ThePRNumber`` with the **pull request number**
```

    $ python -V
    $ python -m venv .venv
    $ source .venv/bin/activate
    $ pip install -e . -U pip -r tests/requirements/py3.txt -r docs/requirements.txt

Run the tests:

.. console::

    $ ./tests/runtests.py

See :ref:`running-unit-tests` for more information. While tests will all pass
in CI, they may not on every person's local environment. For example, only the
latest point release of Python is supported. It's possible you may encounter
errors using an earlier version of Python.

If you have a small number of failures, it may be worth investigating whether
they are known issues, or using the `django-docker-box`_ project to run the
tests. Since this is a sprint, it may be more valuable for you to work on the
issue at hand rather than investigating a weird machine configuration problem.

Build the documentation:

.. console::

    $ cd docs && make html && cd ..

.. _django-docker-box: https://github.com/django/django-docker-box/

Finding a ticket to work on
===========================

Finding a workable ticket is a challenging task. It's important
to read this whole section before attempting to find a ticket to work on.

Assessing tickets
-----------------

> When `looking over tickets`_, it's a good idea to keep your search to

Reviewer: The link to Trac contains the filter "triage stage is accepted". It may be helpful to explicitly mention in the text, too, that only tickets with this triage stage are ready to be worked on.

Author: Rephrased this to  "accepted tickets" in https://github.com/django/django/pull/21338/commits/d833d7ed39c85db5451ff0a606bec482e6055624

consider filtering down to the forms components. Similarly, if you've
never worked with geospatial projects, it's probably wise to avoid
``django.contrib.gis``.

.. _looking over tickets: https://code.djangoproject.com/query?stage=Accepted&status=assigned&status=new&order=id&desc=1&col=id&col=summary&col=owner&col=type&col=component&col=time&col=changetime

Methods to find tickets
-----------------------

Some ways to find a ticket to work on are the following:

* Look at the `easy pickings`_.

* Use the `vulture method`_.

The vulture method is the process where you look for accepted tickets that
haven't been touched in six months or longer. The ideal ticket has a PR
attached to it that has a code review with explicit requests and the author
has not responded in at least six months on the PR and on the ticket.

When you find a stale ticket, make yourself the owner on the ticket with
the "Modify Ticket" button near the bottom of the page. Claim the ticket
by clicking the "assign to" radio button in the "Action" section. Your
username will be filled in the text box by default.

.. _easy pickings: https://code.djangoproject.com/query?status=!closed&easy=1

> .. _vulture method: https://youtube.com/shorts/D6QHet5U82U?si=j5M6sy0ufpeGy_iZ

Reviewer: Maybe a link to Sarah's longer video on the vulture method would be helpful, too: https://youtu.be/A-3eTMNQ3rM?si=DWhLKpFBuEb8FawZ&t=761

And/or a listing of the Trac filters:
Triage stage = Accepted
Has patch = Yes
Patch needs improvement = Yes
Created = between 5 years ago and today
Modified = between [empty] and 6 months ago
Type = Bug

Author: I kind of like the short because it's explicitly short and there isn't a question of, "when do I stop watching this video?". I'm good adding a list of the flags needed, though I'm likely to leave off type and has patch because I think they're redundant.

Author: Added the trac filters here: https://github.com/django/django/pull/21338/commits/d833d7ed39c85db5451ff0a606bec482e6055624 I kept the short video because it's shorter. I'm happy to be corrected on that though.

=======================

> Sprint Quickstart Guide

Reviewer 3: Our documented guide for title requires "In section titles, capitalize only initial words and proper nouns." so I'll push a case fix for this.

=======================

This document outlines the concise guidelines for preparing to contribute to
Django. This should be used during sprints or by people who are more familiar
with contributing to Python open source projects.

This guide assumes you're familiar with:

* Git (see :doc:`/internals/contributing/writing-code/working-with-git`)

* Managing multiple versions of Python

* Python dependency management, including virtual environments

* The concept of running tests and building documentation

Set up
======

Fork the repository, clone your fork to your machine and configure a remote
reference to Django's repository.

.. console::

    $ git clone https://github.com/YourGitHubName/django.git django-repo
    $ cd django-repo
    $ git remote add upstream https://github.com/django/django.git

Confirm you're using a
:ref:`compatible version of Python <faq-python-version-support>`, install
the dependencies to run the tests and build the documentation.

.. console::

    $ # This needs to be >= 3.12

>     $ python -V

Reviewer 5: ```suggestion
    $ python3 -V
```
Invoking `python` will fail in a lot of Linux distributions, where usually `python3` is the alias for a Python 3.* specific version. 

It could be a compromise between `python` and `python3.14`

Author: The `python3` syntax is only used 3 times in our documentation, while `python` is used in pretty much all the other places. I'm inclined to leave this as is because it's more consistent with the project.

    $ python -m venv .venv

>     $ source .venv/bin/activate

Reviewer 5: ```suggestion
    $ . .venv/bin/activate
```
I can be wrong but I think `.` is more portable than `source`

Reviewer 3: Right, I'd still keep `source` here. A lone `.` is a single character with no visual weight, and it's easy to miss or lose: it disappears when someone reads the command off a projector, it's easy to drop when retyping by hand, and if it's missed the failure is confusing (the command appears to succeed but the environment isn't activated). `source` is self-documenting by comparison, which matters more on a page aimed at people setting up quickly at a sprint.

OTOH, `intro/contributing.txt` (which is linked from this page) already documents the dot as the fallback for shells where `source` isn't available, and this page links there, so readers who need it will find it.

Author: We make quite a few assumptions for people with this document. If `.` is reliable, maybe we should go with it? If we discover that it's not working after a sprint or two and people are tripping up on it, we can change this.

>     $ pip install -e . -U pip -r tests/requirements/py3.txt -r docs/requirements.txt

Reviewer 5: ```suggestion
    $ python -m pip install -e . -U pip -r tests/requirements/py3.txt -r docs/requirements.txt
```
I think we agreed in another comment to use pip as a module

Run the tests:

.. console::

    $ ./tests/runtests.py

See :ref:`running-unit-tests` for more information. While tests will all pass
in CI, they may not on every person's local environment. For example, only the
latest point release of Python is supported. It's possible you may encounter
errors using an earlier version of Python.

If you have a small number of failures, it may be worth investigating whether
they are known issues, or using the :ref:`django-docker-box
<running-tests-docker-box>` project to run the tests. Since this is a sprint,
it may be more valuable for you to work on the issue at hand rather than

> investigating a weird machine configuration problem.

Reviewer 5: ```suggestion
investigating a weird machine configuration problem. In this case take notes of which tests fails on your machine, so working on your code you can be sure to recognize if your test failures are related or not with your code changes.
```

Reviewer 3: Good idea, this is worth adding. I'll push a fixup.

Build the documentation:

.. console::

    $ cd docs && make html && cd ..

Finding a ticket to work on
===========================

Finding a workable ticket is a challenging task. It's important
to read this whole section before attempting to find a ticket to work on.

Assessing tickets
-----------------

When looking over tickets, it's a good idea to keep your search to
accepted tickets that were created in the last five years. An easy way to tell
is by the ticket number. Avoid tickets older than 10 years old. These are old
for a reason. You're welcome to try, but they will require a high
degree of perseverance. We recommend you start with tickets created in the

> last five years when getting started.

Reviewer 5: ```suggestion
degree of perseverance.
```
Is this sentence duplicated on purpose or is it an error?

Reviewer 3: It's a reinforcement, but I'm pushing now a rewording.

Consider using the component filter option in Trac to limit your search
to areas you have used before. If you've used forms in a few Django apps,
consider filtering down to the forms components. Similarly, if you've
never worked with geospatial projects, it's probably wise to avoid
``django.contrib.gis``.

* View `tickets with created and modified columns`_.

* View `tickets for the documentation component`_.

.. _tickets with created and modified columns: https://code.djangoproject.com/query?stage=Accepted&status=assigned&status=new&order=id&desc=1&col=id&col=summary&col=owner&col=type&col=component&col=time&col=changetime
.. _tickets for the documentation component: https://code.djangoproject.com/query?component=Documentation&stage=Accepted&status=assigned&status=new&order=id&desc=1&col=id&col=summary&col=owner&col=type&col=component&col=time&col=changetime

Methods to find tickets
-----------------------

Some ways to find a ticket to work on are the following:

* Look at the `easy pickings`_.

* Use the `vulture method`_.

The vulture method is the process where you look for accepted tickets that
haven't been touched in six months or longer. The ideal ticket has a PR

> attached to it that has a code review with explicit requests and the author

Reviewer 5: ```suggestion
attached to it that has a code review with explicit requests for changes and the author
```

Reviewer 3: +1

has not responded in at least six months on the PR and on the ticket.

You can view the list of `possibly vulture-able tickets here`_. The filters
that are applied are:

- Triage stage = Accepted
- Patch needs improvement = Yes
- Created = between ``5 years ago`` and ``today``
- Modified = between ``5 years ago`` and ``6 months ago``

.. _easy pickings: https://code.djangoproject.com/query?status=!closed&easy=1
.. _vulture method: https://youtube.com/shorts/D6QHet5U82U?si=j5M6sy0ufpeGy_iZ
.. _possibly vulture-able tickets here: https://code.djangoproject.com/query?changetime=5+years+ago..6+months+ago&description=~&needs_better_patch=1&status=assigned&status=new&time=5+years+ago..today&order=id&desc=1&col=id&col=summary&col=time&col=changetime&col=needs_better_patch&col=owner&col=type&col=component

Claiming a ticket
-----------------

> When you find a ticket, make yourself the owner on the ticket with

Reviewer 5: ```suggestion
When you find a ticket that nobody already claimed, make yourself the owner on the ticket with
```

Author: I think this particular clarification would be confusing since the vulture method mentioned elsewhere in this document tells the person to assign it to themselves.

Reviewer 3: Indeed, we encourage claiming tickets even if they have an owner (assuming proper check for "inactive" was done.

the "Modify Ticket" button near the bottom of the page. Claim the ticket
by clicking the "assign to" radio button in the "Action" section. Your
username will be filled in the text box by default. The :ref:`full
documentation on claiming tickets can be found here.<claiming-tickets>`

Resuming work from someone else's branch
========================================

If you're continuing the work from someone else's pull request on GitHub,
things can be tricky. You need to pull their branch, but then push it to
your GitHub repository fork.

Run the following, but replace ``ThePRNumber`` with the **pull request number**
and ``TheTicketNumber`` with the **Trac ticket number**.

.. console::

    $ git fetch upstream pull/ThePRNumber/head:ticket_TheTicketNumber
    $ git checkout ticket_TheTicketNumber
    $ git push origin ticket_TheTicketNumber

For example, if you're working on PR 8000 which references ticket 123, the
command to run would be:

.. console::

>     $ git fetch upstream pull/8000/head:ticket_123

Reviewer 5: What happens if people copy and paste this line as is in their terminal?

Author: It errors :grin:

Do you have a suggestion on how this can be done better? I believe I stole this approach from another area in our documentation.

Reviewer 3: I'll push a small proposal, also I'd like to minimally include how to work on a new branch that is not vultured so we cover all basis for sprinters.

=======================
Sprint Quickstart Guide
=======================

This document outlines the concise guidelines for preparing to contribute to
Django. This should be used during sprints or by people who are more familiar
with contributing to Python open source projects.

This guide assumes you're familiar with:

* Git (see :doc:`/internals/contributing/writing-code/working-with-git`)

* Managing multiple versions of Python

* Python dependency management, including virtual environments

* The concept of running tests and building documentation

Set up
======

Fork the repository, clone your fork to your machine and configure a remote
reference to Django's repository.

.. console::

    $ git clone https://github.com/YourGitHubName/django.git django-repo
    $ cd django-repo
    $ git remote add upstream https://github.com/django/django.git

Confirm you're using a
:ref:`compatible version of Python <faq-python-version-support>`, install
the dependencies to run the tests and build the documentation.

.. console::

    $ # This needs to be >= 3.12
    $ python -V
    $ python -m venv .venv

>     $ source .venv/bin/activate

Reviewer 3: One thing to watch with `.. console::`: it auto-generates a Windows tab by mechanically swapping `/` for `\`, which produces an invalid command for the activation step. The setup block currently renders as:

```
...\> REM This needs to be >= 3.12
...\> py -V
...\> py -m venv .venv
...\> source .venv\bin\activate
...\> py -m pip install -e . -U pip -r tests\requirements\py3.txt -r docs\requirements.txt
```

That fourth line is wrong twice: `source` is not a windows cmd builtin, and venv puts scripts in `Scripts\`, not `bin\` (says the LLM check). It should be `.venv\Scripts\activate.bat`. The other lines are fine.

Given that `intro/contributing.txt` hits the same limitation and already works around it by dropping out of `.. console::` for activation and adding a `doscon` admonition, and since the page already assumes the reader is comfortable with virtual environments (it's in the list at the top), my suggestion is to drop the activation line from the block and point at the existing cross-platform instructions instead. Keeps it short and avoids shipping a broken Windows command.

Author: @nessita Did you consider reverting back `shell` considering there are many areas where that is used in the docs already? Specifically in `writing-documentation.txt`

Reviewer 3: @tim-schilling I'm open to it! What did you have in mind for Windows?

`code-block:: shell` renders no Windows tab at all, so `writing-documentation.txt` currently leaves Windows readers with nothing for the activation step. I'd rather not repeat the per-OS commands here either, since `intro/contributing.txt` already covers them.

How about we use `shell` and add a single pointer, something like:

(Spoiler alter, I'm making the Set Up section an enumerated list to ensure readers can easily do back and forth between their terminals and the docs web page and not get lost in prose.)

```rst
#. Create and activate a virtual environment, then install the dependencies to
   run the tests and build the documentation. If you are not using a Unix-like
   system, see :ref:`intro-contributing-install-local-copy` for
   platform-specific commands.

   .. code-block:: shell

       $ python -m venv .venv
       $ source .venv/bin/activate
       $ python -m pip install -e . -U pip -r tests/requirements/py3.txt -r docs/requirements.txt
```

That keeps the block short and self-contained for most readers, without duplicating the OS-specific instructions which is my main worry.

Author: Seems fine to me, but I can live with either. I wasn't sure what the difference was between console and shell and since I switched to console due to a request from you, I wasn't sure if you had considered switching back to shell. It also seems like our windows console translator should handle the activate statement properly since that's effectively a standard.

Reviewer 3: Having the console translator handling the activate statement is a great idea! Would you be available to file a ticket? I'll triage.

* The concept of running tests and building documentation

If any of these are unfamiliar, see :doc:`/intro/contributing` for a longer,
step-by-step tutorial.

Set up
======

#. Fork the repository on GitHub, clone your fork to your machine, and
   configure a remote reference to Django's upstream repository:

   .. console::

       $ git clone https://github.com/YourGitHubName/django.git
       $ cd django
       $ git remote add upstream https://github.com/django/django.git

#. Confirm you're using a :ref:`compatible version of Python
   <faq-python-version-support>`, then set up a virtual environment:

   .. code-block:: shell

       $ # This needs to be >= 3.12
       $ python -V
       $ python -m venv ~/.virtualenvs/djangodev
       $ source ~/.virtualenvs/djangodev/bin/activate

   If you are not using a Unix-like system, see the :ref:`platform-specific
   instructions <intro-contributing-install-local-copy>`.

#. Install the dependencies to run the tests and build the documentation:

   .. console::

       $ python -m pip install -U pip
       $ python -m pip install -e . -r tests/requirements/py3.txt -r docs/requirements.txt

> #. Optionally, install the pre-commit hooks so that formatting and linting run

Author: For some reason an optional step in this document feels wrong. If someone has a problem with using pre-commit, they'll know they don't need to wire it up. I think you could roll this up into the previous section.

Edit: I may be over thinking things here.

Author: @nessita Alright, I guess my reasoning here is that a person shouldn't have to think "do I do this or that" in this document. It's meant to get them from zero to working machine as quickly as possible. I feel like optional steps are places where a person needs to slow down and think. Having them think about "what do I want my machine to have set up?" means they aren't thinking about the ticket or the problem at hand.

I think the context in which we're going to use this is a bit different than our usual documentation and we should take advantage of that because it'll mean less documentation which (should) mean less maintenance.

Reviewer 3: I take the point about not making people stop and decide, and I agree that's the right spirit for this page. I'd keep this one separate though, because of what the step does (changes your git hooks) rather than a general preference for optionality. OTOH "Optionally" also tells someone hitting an install failure that they can move on rather than assuming their setup is broken.

Every other command installs into a venv but `pre-commit install` writes a hook into `.git/hooks`, efectively changing the behavior of every `git commit`. If we fold it in, the first time one of the checks fails the person sees their commit rejected by a tool they (may) don't know they installed.

Shall we perhaps remove it altogether? Maybe, originally, you left it out intentionally and I missed that?

EDIT: who's overthinking now? 🤣

Author: I forgot about it honestly.

I understand your point about the git hooks and it being different than the venv, but I still lean towards just making it part of the steps for a person to do. Pre-commit is pretty common nowadays and a person who would know and care about git hooks has likely heard of pre-commit. A person who is unfamiliar with pre-commit is probably also unfamiliar with git hooks which means they'd be unable to weigh the options.

The part about their commit being rejected is a good thing though. They'll need to read the error and try to understand what needs to change. I see that as a win, not a negative experience.

Reviewer 3: I'm not fully convinced, but this isn't worth holding up the PR over. I'll drop "Optionally" so there's no decision left to make. I kept it as its own bullet since it's the one command that touches something outside the virtual environment, but so it reads as a plain step like the rest.

   each supported Python version. It's possible you may encounter errors using
   an earlier point release.

   If you have a small number of failures, it may be worth investigating
   whether they are known issues, or using the :ref:`django-docker-box
   <running-tests-docker-box>` project to run the tests. Since this is a
   sprint, it may be more valuable for you to work on the issue at hand rather
   than investigating a weird machine configuration problem. Make a note of the
   failing tests before you start, so you can tell whether a later failure was
   caused by your changes.

#. Build the documentation:

   .. console::

       $ cd docs && make html && cd ..

Finding a ticket to work on
===========================

Finding a workable ticket is a challenging task. It's important to read this
section before attempting to do so.

Assessing tickets
-----------------

When looking over tickets, it's a good idea to keep your search to `accepted
tickets created in the last five years`_. Older tickets are old for a reason:
they are usually still open because they are hard, contentious, or both.
You're welcome to try one, but expect it to require a high degree of

> perseverance. Tickets more than 10 years old are best avoided altogether.

Author: ```suggestion
perseverance.
```

The adjusted language lost the nuance that explained why they should be avoided. With this sentence as a standalone point, it seems out of place or even a little antagonistic towards our triaging process. It's probably better to remove it entirely.

Reviewer 3: Yeah, I see your point, applying. Also the link now takes the reader to at most 5 years old tickets so I think they won't end up with a 10 years old ticket.

never worked with geospatial projects, it's probably wise to avoid
``django.contrib.gis``.

* View `tickets with created and modified columns`_.

* View `tickets for the documentation component`_.

.. _tickets with created and modified columns: https://code.djangoproject.com/query?stage=Accepted&status=assigned&status=new&order=id&desc=1&col=id&col=summary&col=owner&col=type&col=component&col=time&col=changetime

> .. _tickets for the documentation component: https://code.djangoproject.com/query?component=Documentation&stage=Accepted&status=assigned&status=new&order=id&desc=1&col=id&col=summary&col=owner&col=type&col=component&col=time&col=changetime

Reviewer: I think these links will be very helpful for people when looking for tickets.

attached to it that has a code review with explicit requests and the author
has not responded in at least six months on the PR and on the ticket.

To view a list of possibly vulture-able tickets, you can run the following
command to generate a link to the page.

.. console::

>     $ python -c "from datetime import date,timedelta;from urllib.parse import urlencode;t=date.today();f=lambda d:(lambda x:f'{x:%b} {x.day}, {x:%Y}')(t+timedelta(d));print('https://code.djangoproject.com/query?'+urlencode([('changetime', f(-1610)+'..'+f(-151)),('description','~'),('needs_better_patch','1'),('status','assigned'),('status','new'),('time' ,f(-1610)+'..'+f(0)),('order','id'),('desc','1'),('col','id'),('col','summary'),('col','time'),('col','changetime'), ('col','needs_better_patch'),('col','owner'),('col','type'),('col','component')]))"

Reviewer: This, too, looks very helpful and provides lots of orientation.

Reviewer 4: Trac supports relative timestamps in natural language, so my hope would be we can just hard-code this link and use times like "6 months ago" [example](https://code.djangoproject.com/query?changetime=6+months+ago..3+months+ago&description=~&status=assigned&status=new&order=id&desc=1&col=id&col=summary&col=owner&col=type&col=component)

Author: Switched to relative timestamps in https://github.com/django/django/pull/21338/commits/d833d7ed39c85db5451ff0a606bec482e6055624

Reviewer: Great that Tract supports relative timestamps in natural language. This looks a lot more readable! Even if the link stopped working for some reason, users will still be able to reproduce the query.
